CMakeLists.txt

```diff
@@ -6,25 +6,43 @@ project (Beast)
 
 set_property (GLOBAL PROPERTY USE_FOLDERS ON)
 
+set(Boost_USE_STATIC_LIBS ON)
+set(Boost_USE_MULTITHREADED ON)
+
+
 if (MSVC)
     set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /MP /W4 /wd4100 /bigobj /D _WIN32_WINNT=0x0601 /D_SCL_SECURE_NO_WARNINGS=1 /D_CRT_SECURE_NO_WARNINGS=1")
     set (CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} /SAFESEH:NO")
-else()
-    set(Boost_USE_STATIC_LIBS ON)
-    set(Boost_USE_MULTITHREADED ON)
-    find_package(Boost REQUIRED COMPONENTS coroutine context thread filesystem program_options system)
-    include_directories(${Boost_INCLUDE_DIRS})
-    link_directories(${Boost_LIBRARY_DIR})
 
+    # Workaround for MSVC having two boost versions - x86 and x64 on same PC in stage folders
+    if ((NOT DEFINED BOOST_ROOT) AND (DEFINED ENV{BOOST_ROOT}))
+        if(CMAKE_SIZEOF_VOID_P EQUAL 8)
+            set(Boost_LIBRARY_DIR $ENV{BOOST_ROOT}/stage64/lib)
+        elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
+            set(Boost_LIBRARY_DIR $ENV{BOOST_ROOT}/stage/lib)
+        endif()
+    elseif(DEFINED BOOST_ROOT)
+        if(CMAKE_SIZEOF_VOID_P EQUAL 8)
+            set(Boost_LIBRARY_DIR ${BOOST_ROOT}/stage64/lib)
+        elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
+            set(Boost_LIBRARY_DIR ${BOOST_ROOT}/stage/lib)
+        endif()
+    endif()
+else()
     set(THREADS_PREFER_PTHREAD_FLAG ON)
     find_package(Threads)
 
-    set(CMAKE_CXX_FLAGS
-      "${CMAKE_CXX_FLAGS} -std=c++11 -Wall -Wpedantic")
+    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=c++11 -Wall -Wpedantic")
 endif()
 
-if (MINGW)
+find_package(Boost REQUIRED COMPONENTS coroutine context thread filesystem program_options system)
+include_directories(${Boost_INCLUDE_DIRS})
+link_directories(${Boost_LIBRARY_DIRS})
```

Comment:
These Boost changes are causing my build all kinds of linker problems. Here are the first few errors:
```
$ cmake --build .
libboost_filesystem-vc140-mt-s-1_62.lib(path_traits.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in main.obj [C:\Dev\Beast\bin64\test\http\bench-tests.vcxproj]
libboost_filesystem-vc140-mt-s-1_62.lib(path_traits.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MDd_DynamicDebug' in main.obj [C:\Dev\Beast\bin64\test\http\bench-tests.vcxproj]
libboost_filesystem-vc140-mt-s-1_62.lib(path.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in main.obj [C:\Dev\Beast\bin64\test\http\bench-tests.vcxproj]
libboost_filesystem-vc140-mt-s-1_62.lib(path.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MDd_DynamicDebug' in main.obj [C:\Dev\Beast\bin64\test\http\bench-tests.vcxproj]
libboost_filesystem-vc140-mt-s-1_62.lib(codecvt_error_category.obj) : error LNK2038: mismatch detected for '_ITERATOR_DEBUG_LEVEL': value '0' doesn't match value '2' in main.obj [C:\Dev\Beast\bin64\test\http\bench-tests.vcxproj]
libboost_filesystem-vc140-mt-s-1_62.lib(codecvt_error_category.obj) : error LNK2038: mismatch detected for 'RuntimeLibrary': value 'MT_StaticRelease' doesn't match value 'MDd_DynamicDebug' in main.obj [C:\Dev\Beast\bin64\test\http\bench-tests.vcxproj]
```
I assume there's some difference between our build environments. To build boost, I unpacked the boost 1.62 tar file into `$BOOST_ROOT`, ran `bootstrap.bat`, then `./bjam.exe --toolset=msvc-14.0 address-model=64 stage`. That left me with the `stage` and `stage64` directories that this seems to be expecting.

I have VS 14.0.25431.01 Update 3 and cmake 3.6.1 installed. I added some messages to the CMakeLists.txt to dump the Boost result variables:
```
$ cmake -G"Visual Studio 14 2015 Win64" -DCMAKE_INSTALL_PREFIX=c:\dev\lib ..    
[...]
CMake Warning at C:/Program Files/CMake/share/cmake-3.6/Modules/FindBoost.cmake:743 (message):
  Imported targets not available for Boost version 106200
Call Stack (most recent call first):
  C:/Program Files/CMake/share/cmake-3.6/Modules/FindBoost.cmake:842 (_Boost_COMPONENT_DEPENDENCIES)
  C:/Program Files/CMake/share/cmake-3.6/Modules/FindBoost.cmake:1395 (_Boost_MISSING_DEPENDENCIES)
  CMakeLists.txt:38 (find_package)
[...]
Boost_INCLUDE_DIRS: C:/boost_1_62_0
Boost_LIBRARY_DIRS: C:/boost_1_62_0//stage64/lib
Boost_LIBRARIES: C:/boost_1_62_0/stage64/lib/libboost_coroutine-vc140-mt-s-1_62.lib;C:/boost_1_62_0/stage64/lib/libboost_context-vc140-mt-s-1_62.lib;C:/boost_1_62_0/stage64/lib/libboost_thread-vc140-mt-s-1_62.lib;C:/boost_1_62_0/stage64/lib/libboost_filesystem-vc140-mt-s-1_62.lib;C:/boost_1_62_0/stage64/lib/libboost_program_options-vc140-mt-s-1_62.lib;C:/boost_1_62_0/stage64/lib/libboost_system-vc140-mt-s-1_62.lib
```
I don't see anything obviously incorrect there, though the warning raises flags.